wdl-lint/src/rules/nonmatching_output.rs

```diff
@@ -0,0 +1,349 @@
+    let mut exact_match = true;
+    // Check for expected entries missing from `meta.outputs`.
+    for (name, span) in &rule.output_keys {
+        if !rule.meta_outputs_keys.contains_key(name) {
+            exact_match = false;
+            if rule.current_meta_span.is_some() {
+                state.add(nonmatching_output(
+                    *span,
+                    name,
+                    rule.name.as_deref().expect("should have a name"),
+                    rule.ty.expect("should have a type"),
+                ));
+            }
+        }
+    }
+
+    // Check for extra entries in `meta.outputs`.
+    for (name, span) in &rule.meta_outputs_keys {
+        if !rule.output_keys.contains_key(name) {
+            exact_match = false;
+            if rule.current_output_span.is_some() {
+                state.add(extra_output_in_meta(
+                    *span,
+                    name,
+                    rule.name.as_deref().expect("should have a name"),
+                    rule.ty.expect("should have a type"),
+                ));
+            }
+        }
+    }
+
+    // Check for out-of-order entries.
+    if exact_match && !rule.meta_outputs_keys.keys().eq(rule.output_keys.keys()) {
+        state.add(out_of_order(
+            rule.current_meta_outputs_span
+                .expect("should have a `meta.outputs` span"),
+            rule.current_output_span
+                .expect("should have an `output` span"),
+            rule.name.as_deref().expect("should have a name"),
+            rule.ty.expect("should have a type"),
+        ));
+    }
+}
+
+impl<'a> Visitor for NonmatchingOutputRule<'a> {
+    type State = Diagnostics;
+
+    fn document(&mut self, _: &mut Self::State, reason: VisitReason, _: &Document) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn workflow_definition(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        workflow: &WorkflowDefinition,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.name = Some(workflow.name().as_str().to_string());
+                self.ty = Some("workflow");
+            }
+            VisitReason::Exit => {
+                if self.current_meta_span.is_some()
+                    && self.current_meta_outputs_span.is_none()
+                    && !self.output_keys.is_empty()
+                {
+                    state.add(missing_outputs_in_meta(
+                        self.current_meta_span.expect("should have a `meta` span"),
+                        self.name.as_deref().expect("should have a name"),
+                        self.ty.expect("should have a type"),
+                    ));
+                } else {
+                    check_matching(state, self);
+                }
+
+                self.name = None;
```

Comment:
Since this is common with the code in the `task_definition` callback, I might pull it out into its own function.